### PR TITLE
typing the debugger

### PR DESCRIPTION
**Pull request by jasonLaster**
Hey guys, mind giving me some feedback on this early draft

### CODE REVIEW

File: _posts/2017-1-20-typing-the-debugger.markdown [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+We switched to Flow in July because we were running into issues incorporating Tcomb and immutable.js [1](https://github.com/devtools-html/debugger.html/pull/325) . This is mostly incidental, but one benefit of having a static type checker is that it simplifies the way the data is represented at runtime. You don’t necessarily have to pass around tcomb objects. Also, after adding types twice, once for tcomb and once for flow, I can safely say that you should budget *atleast* two days to get everything working out alright.
+
+Where initially, **tcomb** was primarily used for our client layer, we were now able to add types to our actions and reducers, which was really nice.
+
+We started representing our Debugger actions in terms of unions:
+
+```js
+type SourceAction =
+  { type: "TOGGLE_PRETTY_PRINT", source: Source }
+  | { type: "SELECT_SOURCE", source: Source, line?: number }
```
Comment (jamiebuilds) [inline _posts/2017-1-20-typing-the-debugger.markdown:84]: Fun fact: you can have a leading `|` for alignment

```js
type SourceAction =
  | { type: "TOGGLE_PRETTY_PRINT", source: Source }
  | { type: "SELECT_SOURCE", source: Source, line?: number }
```

File: _posts/2017-1-20-typing-the-debugger.markdown [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+**Step 2: It’s about the data stupid**
+
+We chose to use  **TComb** for typing the Application in May. We chose it because we wanted a runtime type checker. At the time, we were not sure if would have a build step and using a runtime type checker meant that we would not be changing the syntax of the code. I also believed at the time that a runtime type checker would catch more issues. I was significantly underestimated how awesome **Flow** is.
```
Comment (jamiebuilds) [inline _posts/2017-1-20-typing-the-debugger.markdown]: I ~~was~~ significantly

Comment (jamiebuilds) [reply _posts/2017-1-20-typing-the-debugger.markdown]: There's also a babel plugin for converting flow types to tcomb btw: https://github.com/gcanti/babel-plugin-tcomb, there's also flow-runtime: https://codemix.github.io/flow-runtime/#/

Comment (jasonLaster) [reply _posts/2017-1-20-typing-the-debugger.markdown]: Neat. Didn't know that

File: _posts/2017-1-20-typing-the-debugger.markdown [hunk 3]
```diff
@@ -0,0 +1,159 @@
+---
+layout: post
+title:  "How the Debugger got into the flow"
+date:   2017-1-20
+categories: DevTools JS
+---
+
+Last spring we kicked off a re-write of the Debugger frontend with the goal of being as developer friendly as possible.
```
Comment (clarkbw) [inline _posts/2017-1-20-typing-the-debugger.markdown]: I link to this article when referencing the re-write https://medium.com/mozilla-tech/introducing-debugger-html-8629b943dcf6#.kwp9uxevn

File: _posts/2017-1-20-typing-the-debugger.markdown [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+[Fernando][fernando] added types to our Pause action/reducer in November. I cannot overstate how important this is,  the debugger is *all about* showing pause data and it is complicated.
+
+[Giorgio][giorgio], [Bryan][bryan], and I worked on adding types to our [client][client-types]. These types landed in January, but were complicated enough to be a month in the making. Once they landed [Arthur][arthur] did the really important work of using the client types in the Debugger. This was the first time that we wrote types in one npm package and used them in another, which was pretty scary at the time.
+
+### Conclusion
+
+Adding type checking last summer was one of the better decisions we made. It continues to help us write safer code and adds an additional automated check  during code reviews.
+
+Like many projects, it has been a team effort and to get to where we are now.  Thanks, [James][james], [Bryan][bryan], [Sankha][sankha], [Fernando][fernando], [Zacquary][zackary], [Arthur][arthur]!
```
Comment (hzoo) [inline _posts/2017-1-20-typing-the-debugger.markdown:115]: showing up as ` [Fernando][fernando]` on the site

**Pull request merged**